### PR TITLE
Add unit tests for pkg/kubeapiserver/admission

### PR DESCRIPTION
**Pull request by s-ito-ts**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This adds unit tests to the following file:
```
pkg/kubeapiserver/admission/config.go
```

**Which issue(s) this PR fixes**:
Part of #97355

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/kubeapiserver/admission/config_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	kubePluginInitializer := NewPluginInitializer(
+		cloudConfig,
+		discoveryRESTMapper,
+		quotainstall.NewQuotaConfigurationForAdmission(),
+	)
+
+	testcases := []struct {
+		description                string
+		config                     *Config
+		proxyTransport             *http.Transport
```
Comment (oomichi) [inline pkg/kubeapiserver/admission/config_test.go:86]: This proxyTransport values are the same on all 2 test cases.
Why do we need to provide this as a test parameter?
Do you have a plan to add more test cases and use this parameter on these test cases?

If removing the fixed values from test cases on the code, I feel we can understand the test case purpose more easily.

File: pkg/kubeapiserver/admission/config_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		egressSelector             *egressselector.EgressSelector
```
Comment (oomichi) [inline pkg/kubeapiserver/admission/config_test.go:87]: ditto: The egressSelector values are the same on all 2 test cases.

File: pkg/kubeapiserver/admission/config_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		serviceResolver            webhook.ServiceResolver
```
Comment (oomichi) [inline pkg/kubeapiserver/admission/config_test.go:88]: ditto: The serviceResolver  values are the same on all 2 test cases.

**Review by oomichi — COMMENTED** [review]
Thanks for doing this.
Some comments here, could you check these?

/cc @oomichi

File: pkg/kubeapiserver/admission/config_test.go (hunk 1 above)
Comment (s-ito-ts) [reply pkg/kubeapiserver/admission/config_test.go:86]: Thanks for your review.
proxyTransport and egressSelector, serviceResolver are fixed values.
I will remove these values from the test parameter.

File: pkg/kubeapiserver/admission/config_test.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+type fakeServiceResolver struct{}
+
+func (f *fakeServiceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+	return nil, nil
+}
+
+func TestConfigNew(t *testing.T) {
+
```
[the remark this answers is not in the document]
Comment (s-ito-ts) [reply pkg/kubeapiserver/admission/config_test.go]: Thanks for your review.
I removed it.

File: pkg/kubeapiserver/admission/config_test.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		cmpopts.IgnoreFields(webhookinit.PluginInitializer{}, "authenticationInfoResolverWrapper"),
+		cmp.AllowUnexported(PluginInitializer{}),
+		// restMapper ignores because the order of data changes and the comparison fails.
+		// quotaConfiguration ignores because it is an interface and cannot be compared.
+		cmpopts.IgnoreFields(PluginInitializer{}, "restMapper", "quotaConfiguration"),
+	}
+
+	for _, tc := range testcases {
+
+		pluginInitializers, _, err := tc.config.New(tc.proxyTransport, tc.egressSelector, tc.serviceResolver)
```
[the remark this answers is not in the document]
Comment (s-ito-ts) [reply pkg/kubeapiserver/admission/config_test.go]: I changed to use subtest.

**Comment by oomichi** [comment]
Thanks for updating.

/lgtm

**Comment by s-ito-ts** [comment]
/assign @lavalamp
Could you approve this PR?

**Comment by lavalamp** [comment]
I guess I don't see what value this test is adding? What mistake will it defend against in the future? Right now it looks like it's just a complicated way of testing the QPS configuration?

**Comment by oomichi** [comment]
> 
> 
> I guess I don't see what value this test is adding? What mistake will it defend against in the future? Right now it looks like it's just a complicated way of testing the QPS configuration?

@lavalamp Are you saying it is not worth adding unit tests for this pkg/kubeapiserver/admission/config.go only?
Or it is not worth adding unit tests for any files?

If saying the first one, I can see your point.
We would be able to avoid bugs related to the additional test covered code in the future.

During code review process, it is hard to ask bug-fix code authors to write unit tests related to the PR if we don't have the corresponding unit tests for fixing package file at all at the time. It would be nice to have unit tests as possible for complicated functions which could contain bugs to avoid such situation.

**Comment by lavalamp** [comment]
Unit tests are worth it when the expected value from prevented bugs is greater than the expected cost of maintenance.

This particular test doesn't seem to add coverage for anything that wouldn't be tested already in some other way.

In particular the bugs here are likely to be in how layers are glued together, and this style of test doesn't really exercise that aspect of the production code.

So, it looks to me that this test will add more maintenance cost than bug reduction.

**Comment by s-ito-ts** [comment]
I understand your point.
I close this PR.

/close

**Pull request closed without merging**